File: harness/scripts/final_verify.lua
-- final_verify.lua — capstone gate (single successor: DONE). The decisive checks are
-- recomputed IN-GATE by final_verify.py: full tests/tier_a suite green, all 18 evidence
-- files present and stage-tagged, all 6 [tier-a][stageN] commits present (gptr-mcp repo
-- for stages 5-6), tracked trees clean.
local L = dofile(gralph.profile_dir .. "/scripts/lib.lua")
local p = io.popen("python scripts/final_verify.py")
if not p then gralph.fail("could not launch final_verify.py"); return end
local out = p:read("*a"); p:close()
if not out or out == "" then
  gralph.fail("final_verify.py produced no output — run `python scripts/final_verify.py` manually and fix what it reports")
  return
end
out = out:gsub("%s+", " ")
local function n(k) return L.num(out, k .. "=(%d+)") end

local col = n("suite_collected")
if not col or col == 0 then gralph.fail("full tier_a suite collected 0 tests — a hollow zero cannot finish the build"); return end
if n("suite_errors") ~= 0 then gralph.fail("full tier_a suite has collection errors — some stage's tests no longer import"); return end
if n("suite_failed") ~= 0 then gralph.fail("full tier_a suite not green — fix the implementation, never the tests"); return end
if n("suite_skipped") ~= 0 then gralph.fail("full tier_a suite has skipped tests — skipping is not passing"); return end
if n("evidence_ok") ~= 1 then
  local why = out:match("evidence_reason=([^|]+)") or "an evidence file is missing or mistagged"
  gralph.fail("evidence incomplete: " .. why .. " — every stage needs red/green/smoke evidence produced by the scripts")
  return
end
if n("commits_ok") ~= 1 then
  local why = out:match("commit_reason=([^|]+)") or "a stage commit is missing or a tree is dirty"
  gralph.fail("commit check failed: " .. why)
  return
end

File: harness/scripts/final_verify.py
"""Capstone verifier — run IN-GATE by final_verify.lua.

    python scripts/final_verify.py

Recomputes: full tests/tier_a suite (junit-derived), presence + stage-tag of all 18
evidence files, and all 6 stage commits (both repos where applicable).
Prints ONE machine line; writes no_read/evidence/final.json.
"""
from __future__ import annotations

import json

import check_commit
import hconf


def main() -> int:
    suite = hconf.run_pytest(hconf.TESTS, hconf.EVID / "final_suite.xml")

    evidence_ok, evidence_reason = 1, "-"
    for n in range(1, 7):
        for kind in ("red", "green", "smoke"):
            p = hconf.EVID / f"stage{n}_{kind}.json"
            if not p.exists():
                evidence_ok, evidence_reason = 0, f"missing:stage{n}_{kind}.json"
                break
            try:
                if json.loads(p.read_text(encoding="utf-8")).get("stage") != n:
                    evidence_ok, evidence_reason = 0, f"mistagged:stage{n}_{kind}.json"
                    break
            except json.JSONDecodeError:
                evidence_ok, evidence_reason = 0, f"unparseable:stage{n}_{kind}.json"
                break
        if not evidence_ok:
            break

    commits_ok, commit_reason = 1, "-"
    for n in range(1, 7):
        ok, reason = check_commit.check(n)
        if not ok:
            commits_ok, commit_reason = 0, f"stage{n}: {reason}"
            break

    vals = {f"suite_{k}": v for k, v in suite.items()}
    vals.update({"evidence_ok": evidence_ok, "commits_ok": commits_ok})
    hconf.write_json(hconf.EVID / "final.json",
                     {**vals, "evidence_reason": evidence_reason, "commit_reason": commit_reason})
    line = " ".join(f"{k}={v}" for k, v in vals.items())
    print(f"{line} evidence_reason={evidence_reason}| commit_reason={commit_reason}|")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
